### PR TITLE
UGNI: split files into new sub directories, create common iface class

### PR DESCRIPTION
**Pull request by MattBBaker**
This splits the UGNI code into two folders, base and rdma. RDMA should be full of RDMA specific code, while base contains common code that should be helpful in implementing data grams and smsg.

### CODE REVIEW

**Comment by MattBBaker** [comment]
@shamisp @yosefe please review.

**Comment by shamisp** [comment]
autobot build

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/747/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/184/ for details. (Mellanox internal link)

File: src/uct/ugni/base/ugni_iface.c [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+    return 1;
+}
+
+static ucs_status_t get_cookie(uint32_t *cookie)
+{
+    char           *cookie_str;
+    char           *cookie_token;
+
+    cookie_str = getenv("PMI_GNI_COOKIE");
```
Comment (miked-mellanox) [inline src/uct/ugni/base/ugni_iface.c]: getenv/setenv is not thread-safe API.
I think we have ucx_getenv/setenv or similar which should be used.

Comment (shamisp) [inline src/uct/ugni/base/ugni_iface.c]: correct, it is not thread safe...

Comment (MattBBaker) [inline src/uct/ugni/base/ugni_iface.c]: Alright, do we want that thread safety issue to be a separate PR or should I make a quick patch on this one?

**Comment by shamisp** [comment]
Please open a bug. I would prefer a separate PR

File: src/uct/ugni/base/ugni_iface.c [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+UCS_CLASS_INIT_FUNC(uct_ugni_iface_t, uct_pd_h pd, uct_worker_h worker,
+                           const char *dev_name, uct_iface_ops_t *uct_ugni_iface_ops,
+                           const uct_iface_config_t *tl_config
+                           UCS_STATS_ARG(ucs_stats_node_t *stats_parent))
+{
+  //uct_ugni_iface_config_t *config = ucs_derived_of(tl_config, uct_ugni_iface_config_t);
+  uct_ugni_device_t *dev;
+  //ucs_status_t rc;
```
Comment (shamisp) [inline src/uct/ugni/base/ugni_iface.c]: no C++ comments please

File: src/uct/ugni/rdma/ugni_rdma_iface.c [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+#include "uct/tl/context.h"
+
+#include <uct/ugni/base/ugni_iface.h>
+#include "ugni_rdma_iface.h"
+#include "ugni_rdma_ep.h"
+/*
+UCS_CLASS_DECLARE(uct_ugni_rdma_iface_t, uct_pd_h pd, uct_worker_h worker,
+                  const char *dev_name, size_t rx_headroom,
+                           const uct_iface_config_t *tl_config)
+*/
```
Comment (shamisp) [inline src/uct/ugni/rdma/ugni_rdma_iface.c]: Can we remove it ?

File: src/uct/ugni/rdma/ugni_rdma_ep.c [hunk 4]
```diff
@@ -607,10 +607,10 @@ static ucs_status_t uct_ugni_ep_get_composed_fma_rdma(uct_ep_h tl_ep, void *buff
     rdma_length = length - iface->config.fma_seg_size;
     fma_length = iface->config.fma_seg_size;
 
-    UCT_TL_IFACE_GET_TX_DESC(&iface->super, iface->free_desc_get_buffer,
+    UCT_TL_IFACE_GET_TX_DESC(&iface->super.super, iface->free_desc_get_buffer,
```
Comment (shamisp) [inline src/uct/ugni/rdma/ugni_rdma_ep.c]: looks a bit ugly super.super. Maybe we should macro it or just define another variable ?

**Comment by shamisp** [comment]
@yosefe - any comments ?

File: src/uct/ugni/rdma/ugni_rdma_ep.c (hunk 4 above)
Comment (MattBBaker) [inline src/uct/ugni/rdma/ugni_rdma_ep.c]: I'm trying to think of a macro scheme that wouldn't be ugly. I would be against changing variable names just to avoid this, otherwise we won't be sure if that parent class was supposed to be called parent, or super, or base, or what.

Comment (shamisp) [inline src/uct/ugni/rdma/ugni_rdma_ep.c]: it is always supper

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/748/ for details.

**Comment by shamisp** [comment]
@MattBBaker - side note, once you have comments on the code, please don't do git --amend. Github does not congest this well...

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/185/ for details. (Mellanox internal link)

**Comment by shamisp** [comment]
@yosefe - any comments ?

**Comment by shamisp** [comment]
I opened a PR with some of the improvements.

File: src/uct/ugni/rdma/ugni_rdma_iface.c [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+    UCT_IFACE_MPOOL_CONFIG_FIELDS("FMA", -1, 0, "fma",
+                                  ucs_offsetof(uct_ugni_iface_config_t, mpool),
+                                  "\nAttention: Setting this param with value != -1 is a dangerous thing\n"
+                                  "and could cause deadlock or performance degradation."),
+
+    {NULL}
+};
+
+static ucs_status_t uct_ugni_query_tl_resources(uct_pd_h pd,
```
Comment (shamisp) [inline src/uct/ugni/rdma/ugni_rdma_iface.c]: The function is expected to be the same for rdma and all the rest. Shell we move it to base ?

File: src/uct/ugni/rdma/ugni_rdma_iface.c [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                          ucs_mpool_hugetlb_malloc,     /* allocation hooks */
+                          ucs_mpool_hugetlb_free,       /* free hook */
+                          uct_ugni_base_desc_init,      /* init func */
+                          NULL , &self->free_desc_get);
+    if (UCS_OK != rc) {
+        ucs_error("Mpool creation failed");
+        goto clean_desc;
+    }
+
+    rc = ucs_mpool_create("UGNI-DESC-BUFFER",
```
Comment (shamisp) [inline src/uct/ugni/rdma/ugni_rdma_iface.c]: Not directly related to this change, but I want to raise it here. Can we replace ucs_mpool_create() with uct_iface_mpool_create()

File: src/uct/ugni/rdma/ugni_rdma_iface.c [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    ugni_rc = GNI_CqDestroy(self->super.local_cq);
+    if (GNI_RC_SUCCESS != ugni_rc) {
+        ucs_warn("GNI_CqDestroy failed, Error status: %s %d",
+                 gni_err_str[ugni_rc], ugni_rc);
+    }
+    ugni_rc = GNI_CdmDestroy(self->super.cdm_handle);
+    if (GNI_RC_SUCCESS != ugni_rc) {
+        ucs_warn("GNI_CdmDestroy error status: %s (%d)",
+                 gni_err_str[ugni_rc], ugni_rc);
+    }
```
Comment (shamisp) [inline src/uct/ugni/rdma/ugni_rdma_iface.c]: all the super calls - since it is allocated in base shell we split the cleanup to rdma cleanup and base cleanup, where base is responsible to release CQ and CDM handle

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/800/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/228/ for details. (Mellanox internal link)

File: src/uct/ugni/rdma/ugni_rdma_iface.c (hunk 6 above)
Comment (MattBBaker) [inline src/uct/ugni/rdma/ugni_rdma_iface.c]: I just tried this and it works fine. I'll make a separate PR with this change.

Comment (MattBBaker) [inline src/uct/ugni/rdma/ugni_rdma_iface.c]: It is pushed into my repo and I'll make a PR once this is pulled. Trying to make a PR now looks way ugly.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com:8888/jenkins/job/gh-ucx-pr/806/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/234/ for details. (Mellanox internal link)

File: src/uct/ugni/rdma/ugni_rdma_iface.c (hunk 6 above)
Comment (shamisp) [inline src/uct/ugni/rdma/ugni_rdma_iface.c]: sounds good to me. It has to go as a separate PR.  Thanks

**Comment by shamisp** [comment]
:+1:
